### PR TITLE
librbd: remove crypto image dispatch

### PR DESCRIPTION
**Pull request by orozery**
Crypto image dispatch was added to remap image offsets, accounting for a crypto header at the beginning of an rbd image.
Later on, CryptoInterface* was added as a member to ImageCtx, allowing a much more efficient and simple way to remap offsets.
This will improve cpu cycles not just for encrypted images, but for all images, since the remap_extents API is called even for non-encrypted images.

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by orozery** [comment]
@trociny I have another PR I'm working on for a bug fix when using crypto + migration. It will depend on this PR.
Since Ilya is currently out, will it be possible to review and merge?

**Comment by trociny** [comment]
@orozery It is in my TODO. I hope will have time this week.

File: src/librbd/ImageCtx.cc [hunk 1]
```diff
@@ -541,10 +541,12 @@ librados::IoCtx duplicate_io_ctx(librados::IoCtx& io_ctx) {
...
     }
 
-    io::Extents extents = {{raw_size, 0}};
-    io_image_dispatcher->remap_extents(
-            extents, io::IMAGE_EXTENTS_MAP_TYPE_PHYSICAL_TO_LOGICAL);
-    return extents.front().first;
+    auto crypto = get_crypto();
+    if (crypto != nullptr) {
+      raw_size -= crypto->get_data_offset();
+    }
```
[the remark this answers is not in the document]
Comment (trociny) [reply src/librbd/ImageCtx.cc]: And just wonder, don't we want some assert to ensure the returned data offset is smaller than raw size, before extracting it? But it is up to you to decide if need it.

**Review by trociny — APPROVED** [review]
Otherwise LGTM.

**Comment by ideepika** [comment]
@orozery can you take a look
```
2021-11-22T18:28:06.466 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0x74DF00: TestLibRBD_TestEncryptionLUKS2_Test::TestBody() (test_librbd.cc:2311)
2021-11-22T18:28:06.467 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xFD071D: HandleSehExceptionsInMethodIfSupported<testing::Test, void> (gtest.cc:2605)
2021-11-22T18:28:06.467 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xFD071D: void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (gtest.cc:2641)
2021-11-22T18:28:06.467 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xFC458A: Run (gtest.cc:2680)
2021-11-22T18:28:06.467 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xFC458A: testing::Test::Run() (gtest.cc:2670)
2021-11-22T18:28:06.467 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xFC46F4: Run (gtest.cc:2858)
2021-11-22T18:28:06.468 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xFC46F4: testing::TestInfo::Run() (gtest.cc:2831)
2021-11-22T18:28:06.468 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xFC4C50: Run (gtest.cc:3012)
2021-11-22T18:28:06.468 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xFC4C50: testing::TestSuite::Run() (gtest.cc:2991)
2021-11-22T18:28:06.468 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xFC6864: testing::internal::UnitTestImpl::RunAllTests() (gtest.cc:5723)
2021-11-22T18:28:06.468 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xFD0C8D: HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool> (gtest.cc:2605)
2021-11-22T18:28:06.469 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xFD0C8D: bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) (gtest.cc:2641)
2021-11-22T18:28:06.469 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==
2021-11-22T18:28:06.469 INFO:tasks.workunit.client.0.smithi077.stderr:==39027== 336 (128 direct, 208 indirect) bytes in 2 blocks are definitely lost in loss record 17 of 21
2021-11-22T18:28:06.469 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    at 0x4C35586: operator new(unsigned long) (vg_replace_malloc.c:342)
2021-11-22T18:28:06.469 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xA4D99D: create (BlockCrypto.h:20)
2021-11-22T18:28:06.469 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xA4D99D: librbd::crypto::util::build_crypto(ceph::common::CephContext*, unsigned char const*, unsigned int, unsigned long, unsigned long, boost::intrusive_ptr<librbd::crypto::CryptoInterface>*) (Utils.cc:61)
2021-11-22T18:28:06.470 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xCD59C1: librbd::crypto::luks::LoadRequest<librbd::ImageCtx>::read_volume_key() (LoadRequest.cc:178)
2021-11-22T18:28:06.470 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xCD75BF: librbd::crypto::luks::LoadRequest<librbd::ImageCtx>::handle_read_header(int) (LoadRequest.cc:143)
2021-11-22T18:28:06.470 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0x82461C: Context::complete(int) (Context.h:99)
2021-11-22T18:28:06.470 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0x82D161: void librbd::io::AioCompletion::callback_adapter<Context, &Context::complete>(void*, void*) (AioCompletion.h:81)
2021-11-22T18:28:06.470 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xAF6589: librbd::io::AioCompletion::complete() (AioCompletion.cc:105)
2021-11-22T18:28:06.471 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xAF7A5F: librbd::io::AioCompletion::complete_request(long) (AioCompletion.cc:236)
2021-11-22T18:28:06.471 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xB1902E: librbd::io::ReadResult::C_ObjectReadRequest::finish(int) (ReadResult.cc:207)
2021-11-22T18:28:06.471 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0x82461C: Context::complete(int) (Context.h:99)
2021-11-22T18:28:06.471 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xD69638: librbd::io::ObjectDispatchSpec::C_Dispatcher::finish(int) (ObjectDispatchSpec.cc:33)
2021-11-22T18:28:06.471 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==    by 0xD695A4: librbd::io::ObjectDispatchSpec::C_Dispatcher::complete(int) (ObjectDispatchSpec.cc:23)
2021-11-22T18:28:06.472 INFO:tasks.workunit.client.0.smithi077.stderr:==39027==
2021-11-22T18:28:06.472 INFO:tasks.workunit.client.0.smithi077.stderr:==39027== LEAK SUMMARY:
```
/ceph/teuthology-archive/ideepika-2021-11-22_17:32:37-rbd-wip-deepika2-testing-2021-11-21-2307-distro-basic-smithi/6519673/teuthology.log

**Comment by ideepika** [comment]
@orozery just wanted to check are you getting ping for the this ^

**Comment by orozery** [comment]
@ideepika thanks, I'm working on this.
Indeed this PR has a memory leak which I already prepared a fix for (will push shortly).
Additionally, I found several other leaks already in master, so preparing another PR to fix them as well.

**Comment by ideepika** [comment]
> @ideepika thanks, I'm working on this.
> Indeed this PR has a memory leak which I already prepared a fix for (will push shortly).
> Additionally, I found several other leaks already in master, so preparing another PR to fix them as well.

Great, thanks for the update!

**Comment by orozery** [comment]
@ideepika changes pushed

**Review by ideepika — APPROVED** [review]

**Comment by trociny** [comment]
@orozery It needs rebase after #44114 has been merged.

**Comment by trociny** [comment]
Also, do we want to backport this (to ease further backporting)?

**Comment by orozery** [comment]
@ideepika are we good here?
Same on this:
https://github.com/ceph/ceph/pull/43859

**Review by ideepika — APPROVED** [review]
lgtm

**Comment by orozery** [comment]
@idryomov not sure if github notifies on pushes? but anyway this was rebased

File: src/test/librbd/mock/MockImageCtx.cc [hunk 2]
```diff
@@ -146,4 +146,12 @@ IOContext MockImageCtx::duplicate_data_io_context() {
   return std::make_shared<neorados::IOContext>(*get_data_io_context());
 }
 
+uint64_t MockImageCtx::get_data_offset() const {
+  auto encryption_format_ptr = encryption_format.get();
```
Comment (idryomov) [inline src/test/librbd/mock/MockImageCtx.cc:150]: Same as for the production method:
```
uint64_t MockImageCtx::get_data_offset() const {
  if (encryption_format != nullptr) {
    return encryption_format->get_crypto()->get_data_offset();
  }
  return 0;
}
```

File: src/test/librbd/mock/MockImageCtx.h [hunk 3]
```diff
@@ -237,6 +237,7 @@ struct MockImageCtx {
   ZTracer::Endpoint trace_endpoint;
 
   std::unique_ptr<crypto::MockEncryptionFormat> encryption_format;
+  uint64_t get_data_offset() const;
```
Comment (idryomov) [inline src/test/librbd/mock/MockImageCtx.h:240]: Don't mix method and field declarations -- move it after `duplicate_data_io_context()` declaration.

File: src/librbd/io/Utils.cc [hunk 4]
```diff
@@ -203,24 +200,50 @@ std::pair<Extents, ImageArea> object_to_area_extents(
...
                             ImageArea area) {
-  Extents extents = {{offset, 0}};
-  image_ctx.io_image_dispatcher->remap_to_physical(extents, area);
-  return extents[0].first;
+  switch (area) {
+  case ImageArea::DATA:
+    return offset + image_ctx.get_data_offset();
+  case ImageArea::CRYPTO_HEADER:
+    // direct mapping
+    ceph_assert(image_ctx->get_data_offset() != 0);
```
Comment (idryomov) [inline src/librbd/io/Utils.cc:232]: This doesn't build:
```suggestion
    ceph_assert(image_ctx.get_data_offset() != 0);
```

**Review by idryomov — APPROVED** [review]

**Comment by idryomov** [comment]
No related failures (with Ubuntu 22.04 excluded due to a lab issue):

https://pulpito.ceph.com/dis-2023-09-05_23:49:08-rbd-wip-dis-testing-distro-default-smithi/
https://pulpito.ceph.com/dis-2023-09-06_08:13:54-rbd-wip-dis-testing-distro-default-smithi/
https://pulpito.ceph.com/dis-2023-09-06_14:33:17-rbd-wip-dis-testing-distro-default-smithi/

**Pull request merged**